File: 29-04-2024/Remove every kth node.js
//{ Driver Code Starts
//Initial Template for javascript

"use strict";

process.stdin.resume();
process.stdin.setEncoding("utf-8");

let inputString = "";
let currentLine = 0;



process.stdin.on("data", (inputStdin) => {
  inputString += inputStdin;
});

process.stdin.on("end", (_) => {
  inputString = inputString
    .trim()
    .split("\n")
    .map((string) => {
      return string.trim();
    });

  main();
});

function readLine() {
  return inputString[currentLine++];
}

class Node {
  constructor(x){
    this.data = x;
    this.next = null;
  }
}

function printList(head){
  let s = "";
  while(head){
    s+=head.data;
    s+=" ";
    head = head.next;
  }
  console.log(s);
}

function main() {
  let t = parseInt(readLine());
  let i = 0;
  for (; i < t; i++) {
    let n = parseInt(readLine());
    let arr = readLine().trim().split(" ").map((x) => parseInt(x));
    let k = parseInt(readLine());
    let head = new Node(arr[0]);
    let tail = head;
    for(let j = 1;j<n;j++){
      tail.next = new Node(arr[j]);
      tail = tail.next;
    }
    let obj = new Solution();
    let res = obj.deleteK(head,k);
    printList(res);
  } 

}
// } Driver Code Ends


//User function Template for javascript

/*LINKED LIST NODE
class Node {
  constructor(x){
    this.data = x;
    this.next = null;
  }
}
*/


/**
 * @param {Node} head
 * @param {number} k
 * @return {Node}
*/

class Solution {
  deleteK(head,k){
    //Your Code Start
       if (k === 1) { // If k is 1, every node needs to be deleted, so return null
      return null;
    }

    let dummy = new Node(0); // Create a dummy node to handle edge case when head needs to be deleted
    dummy.next = head;
    let prev = dummy;
    let current = head;
    let count = 0;

    while (current !== null) {
      count++;

      if (count % k === 0) { // If count is a multiple of k, delete the current node
        prev.next = current.next;
      } else {
        prev = current;
      }

      current = current.next;
    }

    return dummy.next;
    //Your Code End
  }
}

File: 29-04-2024/Remove every kth node.cs
//{ Driver Code Starts
//Initial Template for C#

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DriverCode;

namespace DriverCode
{
    public class Node{
        public int data;
        public Node next;
        
        public Node(int a){
            this.data = a;
            this.next = null;
        }
        
    }
    
    class GFG
    {
        static void printList( Node head){
            while(head != null){
                Console.Write(head.data+" ");
                head=head.next;
            }
            Console.Write("\n");
        }
        static void Main(string[] args)
        {
            int T = Convert.ToInt32(Console.ReadLine());
            while(T-- > 0){
                int n = Convert.ToInt32(Console.ReadLine());
                string elements = Console.ReadLine().Trim();
                string []s = elements.Split(' ');
                int[] arr = new int[n];
                arr = Array.ConvertAll(elements.Split(), int.Parse);
                
                Node head = new Node(arr[0]);
                Node tail = head;
                
                for(int i=1; i<n; i++){
                    tail.next = new Node(arr[i]);
                    tail = tail.next;
                }
                
                int k = Convert.ToInt32(Console.ReadLine());
                Solution obj = new Solution();
                Node res = obj.deleteK(head, k);
                printList(res);
                
            }
        }
    }
}
// } Driver Code Ends


//User function Template for C#

/*
public class Node{
    public int data;
    public Node next;

    public Node(int a){
        this.data = a;
        this.next = null;
    }

}
*/
//You Need To Submit this code because in this driver code "The variable `s' is assigned but its value is never used" if you run this code
Compilation succeeded - 1 warning(s)
class Solution
    {
        //Complete this method
        public Node deleteK(Node head, int K)
        {
          //Your Code Start
              if (K == 1) // If K is 1, every node needs to be deleted, so return null
            return null;

        Node dummy = new Node(0); // Create a dummy node to handle edge case when head needs to be deleted
        dummy.next = head;
        Node prev = dummy;
        Node current = head;
        int count = 0;

        while (current != null)
        {
            count++;

            if (count % K == 0) // If count is a multiple of K, delete the current node
            {
                prev.next = current.next;
            }
            else
            {
                prev = current;
            }

            current = current.next;
        }

        return dummy.next;
            //Your code End
        }
    }
